IdentifiabilityODE.mpl

```diff
@@ -557,3 +589,163 @@ CompareDiffVar := proc(dvl, dvr, var_list)
   end if:
   return StringTools[Compare](vr, vl):
 end proc:
+
+
+#===============================================================================
+# Weights
+#===============================================================================
+
+# get rid of unerscore, e.g. x1(t) -> x1_ -> x1
+get_function_name := f -> parse(convert(FunctionToVariable(f), string)[..-2]):
+
+# check if depends on (t)
+is_function:= f->StringTools[Has](convert(f, string), "(t)"):
+idtfm := x->x:
+
+# check if is derivative
+is_diff := f->type(int(f, t), function(name)):
+
+lhs_name := ff -> if convert(ff, string)[-1] = "_" then parse(convert(ff, string)[..-2]) else ff; end if:
+
+#===============================================================================
+GetStateName := proc(state, x_vars, mu)
+#===============================================================================
+  local state_;
+  if state in mu then
+    return state;
+  end if;
+  state_ := parse(cat(StringTools[Join](StringTools[Split](convert(state, string), "_")[..-2], "_"), "_")):
+  if state_ in x_vars then
+    return state_;
+  end if:
+end proc:
+
+#===============================================================================
+GetMinLevelBFS := proc(s, m, x_vars, y_vars, mu, x_eqs, y_eqs_in, all_vars)
+# s: number of states and parameters
+# m: number of outputs
+#===============================================================================
+  # this part is copied from original SIAN code
+  local current_level, visible_states, visibility_table, i, j, continue, poly_d,
+  leader, separant, candidates, each, differentiate_, k, y_eqs:
+  y_eqs := y_eqs_in:
+  current_level := 0:
+  # get functions on level 0, we consider parameters and states indistinguishable
+  # i.e. parameters are states with d/dt = 0
+  visible_states :=  foldl(`union`, op(map(x->indets(rhs(x)) minus {t}, y_eqs))); #select(f -> f in x_zero_vars, ); # map(x->parse(convert(x, string)[..-2]), select(f -> f in x_zero_vars, foldl(`union`, op(map(x->indets(rhs(x)), y_eqs))))):# cat(StringTools[Split](convert(x, string), "_")[1], "_")
+
+  # construct a hash table of "visibility"
+  visibility_table := table([seq(GetStateName(each, x_vars, mu)=current_level, each in visible_states)]):
+  # this is a flag array: if i-th position == 1 then we must differentiat i-th y(t) function 
+  differentiate_ := [seq(1, i=1..nops(y_eqs))]: 
+

```

Comment:
I think we can avoid all the symbolic differentiations in the double for-loop below.
Instead of actual differentiation, we can obtain the variables at level `j` by replacing every variable at level `j - 1` with the variables in the right-hand side of the corresponding equation. This will be much faster and likely shorter.

Reply:
The new version I pushed should resolve the issue.